### PR TITLE
Fkm nonlinear

### PR DESCRIPTION
**Pull request by maierbn**

### CODE REVIEW

File: src/pylife/mesh/gradient.py
Comment (maierbn) [inline src/pylife/mesh/gradient.py:386]: - [ ] Remove `print` statements

**Comment by maierbn** [comment]
- [ ] `process` in `general.py` Revert to previous signature without the flush parameter, move parameter to `fkm_nonlinear.py`
- [ ] Document classes and/or modules with sphinx

**Comment by maierbn** [comment]
@johannes-mueller I cannot push to the fkm_nonlinear branch, do I have all required access rights?

File: src/pylife/stress/rainflow/general.py
Comment (johannes-mueller) [inline src/pylife/stress/rainflow/general.py:456]: Shouldn't this all go to `fkm_nonlinear.py`?

File: src/pylife/stress/rainflow/recorders.py
Comment (johannes-mueller) [inline src/pylife/stress/rainflow/recorders.py:291]: This also can likely be simplified. 

- [ ] Please file issue about refactoring after the merge.

File: tests/materiallaws/test_notch_approximation_law.py
Comment (johannes-mueller) [inline tests/materiallaws/test_notch_approximation_law.py:276]: Consider putting datasets like this into a separate file to make the test more readable.

Like for example https://github.com/boschresearch/pylife/blob/develop/tests/materialdata/woehler/data.py

File: tests/mesh/test_gradient.py
Comment (johannes-mueller) [inline tests/mesh/test_gradient.py:32]: Where are all these whitespaces coming from? Please revert them.

File: tests/mesh/test_gradient.py
Comment (johannes-mueller) [inline tests/mesh/test_gradient.py:263]: Please add a test with the index levels flipped. 

Like this one: https://github.com/boschresearch/pylife/blob/a9fe148ef1b152215a798f683608aca1f095c940/tests/mesh/test_gradient.py#L63

File: tests/stress/rainflow/test_find_turns.py
Comment (johannes-mueller) [inline tests/stress/rainflow/test_find_turns.py:1]: Please revert this change. There should be two blank lines between free functions according to PEP8

**Review by johannes-mueller — CHANGES_REQUESTED** [review]
Thanks @maierbn – this is a lot of good work.

While I see some room for improvement, we should merge it soon to release it as beta to get feedback from users. Then we can rework stuff in smaller steps.

Can you please go through my comments and look if you can apply them.

Thanks

**Comment by maierbn** [comment]
@johannes-mueller I addressed most of the comments and marked them as resolved. Some are not yet resolved, either where I would need your input again or where they contain an issue for after the merge.
I applied two suggestions directly in GitHub and I think they got committed without sign-off. Therefore I had to rebase, add the sign-offs and force push.

**Comment by maierbn** [comment]
@johannes-mueller I fail to install it because of dependency problems:
```
pip install -e .[all]
Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://anu9rng:****@rb-artifactory.bosch.com/artifactory/api/pypi/python-virtual/simple
Obtaining file:///home/mbe7rng/software/pylife
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
ERROR: Could not find a version that satisfies the requirement pandas!=2.1,>=1.4.0 (from pylife[all]) (from versions: 0.1, 0.2, 0.3.0, 0.4.0, 0.4.1, 0.4.2, 0.4.3, 0.5.0, 0.6.0, 0.6.1, 0.7.0, 0.7.1, 0.7.2, 0.7.3, 0.8.0, 0.8.1, 0.9.0, 0.9.1, 0.10.0, 0.10.1, 0.11.0, 0.12.0, 0.13.0, 0.13.1, 0.14.0, 0.14.1, 0.15.0, 0.15.1, 0.15.2, 0.16.0, 0.16.1, 0.16.2, 0.17.0, 0.17.1, 0.18.0, 0.18.1, 0.19.0, 0.19.1, 0.19.2, 0.20.0, 0.20.1, 0.20.2, 0.20.3, 0.21.0, 0.21.1, 0.22.0, 0.23.0, 0.23.1, 0.23.2, 0.23.3, 0.23.4, 0.24.0, 0.24.1, 0.24.2, 0.25.0, 0.25.1, 0.25.2, 0.25.3, 1.0.0, 1.0.1, 1.0.2, 1.0.3, 1.0.4, 1.0.5, 1.1.0, 1.1.1, 1.1.2, 1.1.3, 1.1.4, 1.1.5)
ERROR: No matching distribution found for pandas!=2.1,>=1.4.0
```
Is the spec. of `numpy==1.23.5` really necessary?

**Comment by johannes-mueller** [comment]
> ```
> 
> ```
> 
> 
>     
>   
> 
> Is the spec. of `numpy==1.23.5` really necessary?

It used to be necessary because numba had a problem with newer versions. Maybe they fixed it in the meantime and we can try to uplift the restriction.

I filed #39 in order to check.

**Comment by maierbn** [comment]
Okay, seems like the numpy persion fix is required for now.. I now managed to install it with the restriction. 
Still the docs won't build, I tried the following:
```
python -m pip install -e .[docs]    # was missing jupyter, therefore I added:
python -m pip install -e .[extras]
```
Then in `docs` directory:
```
make html
```
Now it complains about `pandoc`: 
```
$ make html
Sphinx v7.2.6 in Verwendung
making output directory... erledigt
loading intersphinx inventory from http://www.sphinx-doc.org/en/master/objects.inv...
loading intersphinx inventory from https://docs.python.org/3.10/objects.inv...
loading intersphinx inventory from https://matplotlib.org/objects.inv...
loading intersphinx inventory from https://numpy.org/doc/stable/objects.inv...
loading intersphinx inventory from https://scikit-learn.org/stable/objects.inv...
loading intersphinx inventory from https://pandas.pydata.org/pandas-docs/stable/objects.inv...
loading intersphinx inventory from https://docs.scipy.org/doc/scipy/reference/objects.inv...
loading intersphinx inventory from https://pyscaffold.org/en/stable/objects.inv...
intersphinx inventory has moved: http://www.sphinx-doc.org/en/master/objects.inv -> https://www.sphinx-doc.org/en/master/objects.inv
intersphinx inventory has moved: https://matplotlib.org/objects.inv -> https://matplotlib.org/stable/objects.inv
intersphinx inventory has moved: https://docs.scipy.org/doc/scipy/reference/objects.inv -> https://docs.scipy.org/doc/scipy/objects.inv
[autosummary] generating autosummary for: 3rd-party-licenses.md, CHANGELOG.md, CODINGSTYLE.rst, CONTRIBUTING.md, INSTALLATION.md, LICENSE.md, NEWS-2.0.md, NOTICE.md, README.md, broadcaster.rst, ..., tutorials.rst, tutorials/stress-strength.rst, user_guide.rst, utils/functions.rst, utils/histogram.rst, utils/probability_data.rst, variable_names.rst, vmap/vmap.rst, vmap/vmap_export.rst, vmap/vmap_import.rst
WARNING (pytensor.tensor.blas): Using NumPy C-API based implementation for BLAS functions.
myst v2.0.0: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=set(), disable_syntax=[], all_links_external=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, fence_as_directive=set(), number_code_blocks=[], title_to_header=False, heading_anchors=0, heading_slug_func=None, html_meta={}, footnote_transition=True, words_per_minute=200, substitutions={}, linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', enable_checkboxes=False, suppress_warnings=[], highlight_code_blocks=True)
building [mo]: targets for 0 po files that are out of date
writing output... 
building [html]: targets for 79 source files that are out of date
updating environment: [new config] 79 added, 0 changed, 0 removed
executing broadcaster13%] broadcaster
reading sources... [ 16%] demos/hotspot_beam
Notebook error:
PandocMissing in demos/hotspot_beam.nblink:
Pandoc wasn't found.
Please check that pandoc is installed:
https://pandoc.org/installing.html
Makefile:29: recipe for target 'html' failed
make: *** [html] Error 2
```
Installing it `python -m pip install pandoc` does not change anything. How is the documentation normally build? Maybe could we add a Github Action to check if it builds?

**Comment by johannes-mueller** [comment]
> How is the documentation normally build? Maybe could we add a Github Action to check if it builds?

It is normally built by readthedocs fetching the repo and build it. A github action to check if it builds sure would be a good thing, although it might not guarantee that it builds on readthedocs. I will look into this tomorrow.

**Comment by johannes-mueller** [comment]
> Installing it `python -m pip install pandoc` does not change anything. How is the documentation normally build? Maybe could we add a Github Action to check if it builds?

You actually need to install the `pandoc` software itself on your host system. The `pandoc` package from PyPi is only the python bindings for.

In the meantime I merged #39 and #40 to `develop`. Can you please merge `develop` to trigger the renewed CI?

**Comment by maierbn** [comment]
I merged develop and fixed the documentation build.

**Review by johannes-mueller — APPROVED** [review]

**Comment by johannes-mueller** [comment]
Squashed manually, thanks.

File: src/pylife/materiallaws/notch_approximation_law.py
Comment (johannes-mueller) [reply src/pylife/materiallaws/notch_approximation_law.py:81]: @maierbn Is this resolved? Is it correct that `K()` sets `K_prime` and vice verca?

**Pull request closed without merging**
